servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -374,6 +396,38 @@ default HttpServerContext listenBlockingStreamingAndAwait(BlockingStreamingHttpS
         return blockingInvocation(listenBlockingStreaming(service));
     }
 
+    /**
+     * Starts this server and returns the {@link HttpServerContext} after the server has been successfully started.
+     * <p>
+     * If the underlying protocol (e.g. TCP) supports it this will result in a socket bind/listen on {@code address}.
+     * <p>
+     * Note that this method is generic in the sense that it accepts all HTTP {@link Service} implementations to be
+     * passed in, namely {@link StreamingHttpService}, {@link HttpService}, {@link BlockingStreamingHttpService} and
+     * {@link BlockingHttpService}. It is especially useful when Dependency Injection is used and the type of service
+     * is not known at compile time.
+     *
+     * @param service Service invoked for every request received by this server. The returned {@link HttpServerContext}
+     * manages the lifecycle of the {@code service}, ensuring it is closed when the {@link HttpServerContext} is closed.
+     * @return A {@link Single} that completes when the server is successfully started or terminates with an error if
+     * the server could not be started.
+     * @throws IllegalArgumentException if an unknown/unsupported {@link Service} is being provided.
+     */
+    default Single<HttpServerContext> listenService(final Service service) {
+        Objects.requireNonNull(service);
```

Comment:
@idelpivnitskiy should we defer() the whole block or keep it eager as is?

Reply:
No, in this case we don't need to defer bcz the incoming argument is checked once and never changes later, and we don't do any meaningful work here (we only checking types). `defer` is useful when we make some operations and want to re-run those operations on every re-subscribe. Example: logic that changes headers of a request.

Also, `requireNonNull` isn't necessary here, it will just fall under `IllegalArgumentException`.